### ISSUE TITLE
`ghcide` test failures in Gentoo Linux

### ISSUE THREAD

**Issue by hololeap** [haskell/haskell-language-server#3221]
I'm getting a few test failures for `ghcide-1.8.0.0` when they are run with Gentoo's network-sandbox disabled (so unrelated to #3126):

1. ```
   FAIL (31.99s)
        test/exe/Main.hs:1653:
        expected: ["AAAAA"]
         but got: ["AAAAA","AAAA","AAAA"]
        Use -p '/incomplete entries/' to rerun this test only.
   ```

2. ```
   FAIL (33.01s)
        test/exe/Main.hs:1807:
        expected: ["Defined in 'Data.Map","Defined in 'Data.Map.Lazy","Defined in 'Data.Map.Strict"]
         but got: []
        Use -p '/Map/' to rerun this test only.
   ```

3. ```
   FAIL
      Exception: /var/tmp/portage/dev-haskell/ghcide-1.8.0.0/temp/extra-dir-12909185275211/other_loc/Sym.hs: createFileLink:createSymbolicLink: does not exist (No such file or directory)
      Use -p '/Module is symlinked/' to rerun this test only.
   ```

```
3 out of 344 tests failed (2492.75s)
```

---

Full build log is [here](https://gist.github.com/0b435bf37f8f277a408ded5fe2eeb15a) (raw file is ~15MB)

---

### Your environment

GHC 9.0.2 on Gentoo Linux. All packages are installed through the system package manager.

`ghcide-1.8.0.0` is being built with the `ghc-patched-unboxed-bytecode` flag turned OFF. (I hit other test failures when it was turned on.)

---

Other `gentoo-haskell`-related test issues:

- https://github.com/haskell/haskell-language-server/issues/3126
- https://github.com/haskell/haskell-language-server/issues/3217
- https://github.com/haskell/haskell-language-server/issues/3218

**Comment by pepeiborra** [comment]
The problem is that Tasty runs test in parallel by default, and most HLS test suites do not handle this well. Set TASTY_NUM_THREADS to 1

**Comment by pepeiborra** [comment]
> > The problem is that Tasty runs test in parallel by default, and most HLS test suites do not handle this well.
> 
> 
> 
> Isn't the number of threads a tasty option? Could we set it explicitly in the test suites to 1 using `adjustOption`?

That would be great

**Comment by michaelpj** [comment]
I made an issue for this: https://github.com/haskell/haskell-language-server/issues/3245

**Comment by michaelpj** [comment]
Did Pepe's suggestion help?

**Comment by hololeap** [comment]
It seems that the `Map` test fails consistently even with @pepeiborra's suggestion and also using cabal-v1/cabal-v2. The others are not as consistent.

I want to try to get a sandbox set up that mirrors the CI exactly, hopefully get all the tests passing, then work backwards and see what is different in Gentoo compared to the sandbox (package versions, etc.)

**Comment by hololeap** [comment]
I _cannot_ get the `Map` test to pass with `cabal-v1` no matter what I try! Here are the steps I took to verify I could get all tests passing with `cabal-v2` and then set up the same environment for `cabal-v1`:

1. I checked out e36208bf72f14fb5779c7cfc9031ba82b2dc6fd5, as this was the latest commit at the time and all CI tests were passing.
2. 
    ```bash
    cabal v2-configure --enable-tests --disable-benchmarks -O0 --flag=-ekg
    cabal v2-build all
    export TASTY_NUM_THREADS=1
    cabal v2-test ghcide --test-show-details=streaming --test-option=--color=always --test-option=--ansi-tricks=false
    ```

    All tests pass! 🎉
3. I created [`cabal.project.freeze`](https://gist.github.com/0e53ee225801ba253884cdb8e305c119) and made sure _all_ packages installed through Gentoo were consistent with the versions in the file (the only exception was `hsc2hs` which was version `0.68.8` in the freeze file and `0.68.7` on the system.)
4. 
    ```bash
    cd ghcide
    cabal v1-configure --enable-tests --disable-benchmarks -O0 --flag=-ekg
    cabal v1-build
    export PATH="$(realpath dist/build/ghcide):$(realpath dist/build/ghcide-test-preprocessor):$PATH"
    export TASTY_NUM_THREADS=1
    cabal v1-test --show-details=streaming --test-option=--color=always
    ```

    ```text
          Map:                                                                            FAIL (14.60s)
            test/exe/Main.hs:1807:
            expected: ["Defined in 'Data.Map","Defined in 'Data.Map.Lazy","Defined in 'Data.Map.Strict"]
             but got: []
            Use -p '/Map/' to rerun this test only.
    ```

    ```text
    1 out of 344 tests failed (1512.25s)
    ```

    😠

**Comment by hololeap** [comment]
Hm, it looks like I'm getting some inconsistencies with `Map` using cabal-v2 as well:

1.
    ```console
    $ cabal v2-test ghcide --test-show-details=streaming --test-option=--color=always --test-option=--ansi-tricks=false --test-option=-p --test-option='/Map/'
    Build profile: -w ghc-9.0.2 -O0
    In order, the following will be built (use -v for more details):
    - ghcide-1.8.0.0 (test:ghcide-tests) (first run)
    Preprocessing test suite 'ghcide-tests' for ghcide-1.8.0.0..
    Building test suite 'ghcide-tests' for ghcide-1.8.0.0..

    <no location info>: warning: [-Wunused-packages]
        The following packages were specified via -package or -package-id flags,
        but were not needed for compilation:
        - record-hasfield-1.0
        - record-dot-preprocessor-0.2.15
        - ghc-typelits-knownnat-0.7.6
        - ghc-9.0.2

    <no location info>: warning: [-Wunused-packages]
        The following packages were specified via -package or -package-id flags,
        but were not needed for compilation:
        - record-hasfield-1.0
        - record-dot-preprocessor-0.2.15
        - ghc-typelits-knownnat-0.7.6
        - ghc-9.0.2
    Running 1 test suites...
    Test suite ghcide-tests: RUNNING...
    ghcide
    completion
        package
        Map: FAIL (17.58s)
            test/exe/Main.hs:1807:
            expected: ["Defined in 'Data.Map","Defined in 'Data.Map.Lazy","Defined in 'Data.Map.Strict"]
            but got: []

    1 out of 1 tests failed (17.58s)
    Test suite ghcide-tests: FAIL
    Test suite logged to:
    /home/admin/haskell-language-server-cabal/dist-newstyle/build/x86_64-linux/ghc-9.0.2/ghcide-1.8.0.0/t/ghcide-tests/noopt/test/ghcide-1.8.0.0-ghcide-tests.log
    0 of 1 test suites (0 of 1 test cases) passed.
    cabal: Tests failed for test:ghcide-tests from ghcide-1.8.0.0.
    ```

2.
    ```console
    $ cabal v2-test ghcide --test-show-details=streaming --test-option=--color=always --test-option=--ansi-tricks=false --test-option=-p --test-option='/Map/'
    Build profile: -w ghc-9.0.2 -O0
    In order, the following will be built (use -v for more details):
    - ghcide-1.8.0.0 (test:ghcide-tests) (first run)
    Preprocessing test suite 'ghcide-tests' for ghcide-1.8.0.0..
    Building test suite 'ghcide-tests' for ghcide-1.8.0.0..

    <no location info>: warning: [-Wunused-packages]
        The following packages were specified via -package or -package-id flags,
        but were not needed for compilation:
        - record-hasfield-1.0
        - record-dot-preprocessor-0.2.15
        - ghc-typelits-knownnat-0.7.6
        - ghc-9.0.2

    <no location info>: warning: [-Wunused-packages]
        The following packages were specified via -package or -package-id flags,
        but were not needed for compilation:
        - record-hasfield-1.0
        - record-dot-preprocessor-0.2.15
        - ghc-typelits-knownnat-0.7.6
        - ghc-9.0.2
    Running 1 test suites...
    Test suite ghcide-tests: RUNNING...
    ghcide
    completion
        package
        Map: FAIL (17.58s)
            test/exe/Main.hs:1807:
            expected: ["Defined in 'Data.Map","Defined in 'Data.Map.Lazy","Defined in 'Data.Map.Strict"]
            but got: []

    1 out of 1 tests failed (17.58s)
    Test suite ghcide-tests: FAIL
    Test suite logged to:
    /home/admin/haskell-language-server-cabal/dist-newstyle/build/x86_64-linux/ghc-9.0.2/ghcide-1.8.0.0/t/ghcide-tests/noopt/test/ghcide-1.8.0.0-ghcide-tests.log
    0 of 1 test suites (0 of 1 test cases) passed.
    cabal: Tests failed for test:ghcide-tests from ghcide-1.8.0.0.
    ```

It could be relevant that both of these cases take exactly 17.58s before a failure. Is there some kind of hidden timeout occurring?

**Comment by hololeap** [comment]
I'm going to play around with [the `source` vs. `installed` constraints](https://cabal.readthedocs.io/en/3.8/setup-commands.html#cmdoption-runhaskell-Setup.hs-configure-constraint) and see if I can bisect down to a specific set of packages that are introducing the bug when using `installed` versions.
